lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/cache/EvictableLoadingCache.java

```diff
@@ -0,0 +1,369 @@
+        }
+
+        if (refreshMillis.isPresent() && (expiresAfterWriteMillis.isEmpty() || expiresAfterWriteMillis.getAsLong() > refreshMillis.getAsLong())) {
+            dataCache.refreshAfterWrite(refreshMillis.getAsLong(), MILLISECONDS);
+        }
+
+        tokenCache.maximumSize(maximumSize);
+        dataCache.maximumSize(maximumSize);
+
+        if (recordStats) {
+            dataCache.recordStats();
+        }
+
+        return new EvictableLoadingCache<>(
+                EvictableCache.buildWith(tokenCache),
+                dataCache.build(new TokenClassLoader<>(cacheLoader)));
+    }
+
+    private final EvictableCache<K, Token<K>> tokensCache;
+    private final LoadingCache<Token<K>, V> dataCache;
+
+    private EvictableLoadingCache(EvictableCache<K, Token<K>> tokensCache, LoadingCache<Token<K>, V> dataCache)
+    {
+        this.tokensCache = tokensCache;
+        this.dataCache = dataCache;
+    }
+
+    @Override
+    public V get(K key)
+            throws ExecutionException
+    {
+        Token<K> token = tokensCache.get(key, () -> new Token<>(key));
+        return dataCache.get(token);
+    }
+
+    @Override
+    public ImmutableMap<K, V> getAll(Iterable<? extends K> keys)
+            throws ExecutionException
+    {
+        BiMap<K, Token<K>> keyToToken = HashBiMap.create();
+        for (K key : keys) {
+            // This is not bulk, but is fast local operation
+            keyToToken.put(key, tokensCache.get(key, () -> new Token<>(key)));
+        }
+
+        Map<Token<K>, V> values = dataCache.getAll(keyToToken.values());
+
+        BiMap<Token<K>, K> tokenToKey = keyToToken.inverse();
+        ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+        for (Map.Entry<Token<K>, V> entry : values.entrySet()) {
+            Token<K> token = entry.getKey();
+
+            // While token.getKey() returns equal key, a caller may expect us to maintain key identity, in case equal keys are still distinguishable.
+            K key = tokenToKey.get(token);
+            checkState(key != null, "No key found for %s in %s when loading %s", token, tokenToKey, keys);
+
+            result.put(key, entry.getValue());
+        }
+        return result.buildOrThrow();
+    }
+
+    @Override
+    public V getIfPresent(Object key)
+    {
+        Token<K> token = tokensCache.getIfPresent(key);
+        if (token == null) {
+            return null;
+        }
+        return dataCache.getIfPresent(token);
+    }
+
+    @Override
+    public void invalidate(Object key)
+    {
+        Token<K> token = tokensCache.getIfPresent(key);
+        // Invalidate irrespective of whether we got a token, to invalidate result of ongoing token load (instantiation).
+        tokensCache.invalidate(key);
+        if (token != null) {
+            dataCache.invalidate(token);
+        }
+    }
+
+    @Override
+    public void invalidateAll(Iterable<?> keys)
+    {
+        tokensCache.invalidateAll(keys);
```

Comment:
Why do you invalidate before `get*Present` and above after `get*Present`? I would expect here the same. And why invalidate twice?

Reply:
copy&paste error, fixed, thanks.